File: src/main/java/com/joraph/schema/BaseProperty.java
package com.joraph.schema;

import static java.util.Objects.requireNonNull;

import com.joraph.JoraphException;

import kotlin.jvm.functions.Function1;

/**
 * Base class for {@code Property}.
 * @param <T> the property type
 */
public class BaseProperty<T, R>
	implements Property<T, R> {

	private Function1<T, R> accessor;

	/**
	 */
	public BaseProperty() {
	}

	/**
	 * @param accessor the accessor to set
	 */
	public BaseProperty(Function1<T, R> accessor) {
		setPropertyChain(accessor);
	}

	/**
	 */
	protected void setPropertyChain(Function1<T, R> accessor) {
		this.accessor = requireNonNull(accessor, "accessor must not be null");
	}

	/**
	 * @return descriptor the descriptor
	 */
	protected Function1<T, R> getPropertyAccessor() {
		return this.accessor;
	}

	@Override
	@SuppressWarnings("unchecked")
	public R read(Object obj) {
		try {
			return accessor.invoke((T)obj);
		} catch (Exception e) {
			throw new JoraphException(e);
		}
	}

}

File: src/main/kotlin/com/joraph/schema/BaseProperty.kt
package com.joraph.schema

import com.joraph.JoraphException

/**
 * Base class for [Property] that uses a function for accessing the Property's value.
 */
open class BaseProperty<T, R> constructor(val propertyAccessor: (T) -> R?) : Property<T, R> {

    @Suppress("UNCHECKED_CAST")
    override fun read(obj: Any): R? {
        return try {
            propertyAccessor.invoke(obj as T)
        } catch (e: Exception) {
            throw JoraphException(e)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as BaseProperty<*, *>

        if (propertyAccessor != other.propertyAccessor) return false

        return true
    }

    override fun hashCode(): Int {
        return propertyAccessor.hashCode()
    }


}
